### ISSUE TITLE
ParseFromRequestWithClaims with Unique Parser

### ISSUE THREAD

**Issue by jon-whit** [dgrijalva/jwt-go#245]
I'd like to be able to parse a JWT from an HTTP request using the `request` package, but have token validation be performed using only a specific set of ValidMethods. A use case for this may be to not allow an empty "alg" type (e.g., "alg": "").

The issue is that the `request.ParseFromRequestWithClaims` function calls upon the function that creates a new (empty) `jwt.Parser` ([Line 93](https://github.com/dgrijalva/jwt-go/blob/master/token.go#L93) of token.go).

It would be nice to be able to create a Parser, and call upon a method that can parse the request with claims and validate against the slice of ValidationMethods provided to the Parser structure.

For example:

```
p := &Parser {
    ValidationMethods: [ ] string {"RS256"}
    SkipClaimsValidation: false
}

p.ParseFromRequestWithClaims(req, extractor, claims, keyFunc)
```

This way lines such as [these](https://github.com/dgrijalva/jwt-go/blob/master/parser.go#L76-L89) get hit.

Does this support already exist somewhere that I cannot find? If not, I'd like to submit a pull-request to add this functionality. My team needs this quite badly.

**Comment by dgrijalva** [comment]
I think I understand what you're looking for. You can do this by breaking the task into parts:

First, extract the token from the request. The interface `request.Extractor` defines a method you can call directly to use the extractor on its own. You'll then have the token string which you can use with the rest of the library.

Because you have the token string, you can now create your own instance of parser and set the flags however you like.

Additionally, if you need to overwrite the validation behavior, you can do this as well by implementing your own `Claims` type. You can use this method to inherit and override the built in validation behaviors by using go's composition features.

```go
type MyClaims{
   *jwt.StandardClaims
   Foo string `json:"foo,omitempty"`
}

func (c *MyClaims) Valid()error {
  // Utilize built in validations
  if err := c.StandardClaims.Valid(); err != nil {
    return error
  } 

  // your logic here
  

  return nil
} 
```

Finally, if you want to do special things based on the header section of the token, you have two choices:

1. Validate header within your `Keyfunc`
2. Perform custom validations after you finish parsing the token

Between all those techniques, you should be able to get to what you're after. Did this help or did I misunderstand the question?

**Comment by jon-whit** [comment]
@dgrijalva yeah that's perfect. That's exactly what I'm looking for!

It would be nice if that was a method of the Parser structure.It would allow an instantiated Parser structure to parse the token from the request and validate it against the parameters of the Parser.

**Comment by dgrijalva** [comment]
I don't think that's an option as it would create a circular reference. What we can do is add parser as an option to the `ParseFromRequest..` helpers. I'll probably use the option function pattern that's becoming popular to do so.

**Comment by dgrijalva** [comment]
Landed in master for 3.2.0 release

**Comment by jon-whit** [comment]
@dgrijalva that's great! Thanks for adding that :)
